flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java

```diff
@@ -36,27 +40,58 @@
  */
 public class SortPartitionOperator<T> extends SingleInputOperator<T, T, SortPartitionOperator<T>> {
 
-	private int[] sortKeyPositions;
+	private List<Keys<T>> keys;
 
-	private Order[] sortOrders;
+	private List<Order> orders;
 
 	private final String sortLocationName;
 
+	private boolean useKeySelector;
 
-	public SortPartitionOperator(DataSet<T> dataSet, int sortField, Order sortOrder, String sortLocationName) {
+	private SortPartitionOperator(DataSet<T> dataSet, String sortLocationName) {
 		super(dataSet, dataSet.getType());
+
+		keys = new ArrayList<>();
+		orders = new ArrayList<>();
 		this.sortLocationName = sortLocationName;
+	}
+
+
+	public SortPartitionOperator(DataSet<T> dataSet, int sortField, Order sortOrder, String sortLocationName) {
+		this(dataSet, sortLocationName);
+		this.useKeySelector = false;
+
+		ensureSortableKey(sortField);
 
-		int[] flatOrderKeys = getFlatFields(sortField);
-		this.appendSorting(flatOrderKeys, sortOrder);
+		keys.add(new Keys.ExpressionKeys<>(sortField, getType()));
+		orders.add(sortOrder);
 	}
 
 	public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrder, String sortLocationName) {
-		super(dataSet, dataSet.getType());
-		this.sortLocationName = sortLocationName;
+		this(dataSet, sortLocationName);
+		this.useKeySelector = false;
+
+		ensureSortableKey(sortField);
+
+		keys.add(new Keys.ExpressionKeys<>(sortField, getType()));
+		orders.add(sortOrder);
+	}
+
+	public SortPartitionOperator(DataSet<T> dataSet, Keys<T> sortKey, Order sortOrder, String sortLocationName) {
```

Comment:
Change the `sortKey` parameter type to `SelectorFunctionKeys` (or accept the `KeySelector` and create the `SelectorFunctionKeys` in the constructor.